apollo-runtime/src/main/java/com/apollographql/apollo/internal/cache/http/Utils.java

```diff
@@ -84,16 +98,51 @@ static void copyResponseBody(Response response, Sink sink) throws IOException {
     closeQuietly(responseBodySource);
   }
 
-  private static void closeQuietly(Source source) {
+  static void closeQuietly(Source source) {
     try {
       source.close();
     } catch (Exception ignore) {
       // ignore
     }
   }
 
-  private static HttpCacheControl cacheControl(Request request) {
-    return HttpCacheControl.valueOfHttpHeader(request.header(HttpCache.CACHE_CONTROL_HEADER));
+  static void closeQuietly(Response response) {
+    try {
+      if (response != null) {
+        response.close();
+      }
+    } catch (Exception ignore) {
+      // ignore
+    }
+  }
+
+  static boolean isStale(Request request, Response response) {
+    String timeoutStr = request.header(CACHE_EXPIRE_TIMEOUT_HEADER);
+    String servedDateStr = response.header(CACHE_SERVED_DATE_HEADER);
+    if (servedDateStr == null || timeoutStr == null) {
+      return true;
+    }
+
+    long timeout = Long.parseLong(timeoutStr);
+    Date servedDate = HttpDate.parse(servedDateStr);
+
+    long now = System.currentTimeMillis();
+    return servedDate == null || now - servedDate.getTime() > timeout;
+  }
+
+  private static FetchStrategy fetchStrategy(Request request) {
+    String fetchStrategyHeader = request.header(CACHE_FETCH_STRATEGY_HEADER);
+    if (fetchStrategyHeader == null || fetchStrategyHeader.isEmpty()) {
+      return null;
+    }
+
+    for (FetchStrategy fetchStrategy : FetchStrategy.values()) {
```

Comment:
valueof throws exception if it can't find enum const by name.
> IllegalArgumentException - if this enum type has no constant with the specified name

this one is safe